### PR TITLE
Shabab Ali: PySpark Homework Submission

### PR DESCRIPTION
**Pull request by shababmoali**

### CODE REVIEW

**Comment by julie-scherer** [comment]
### Feedback:
### Coverage of Homework Prompt
The student's solution partially addresses the homework prompt. They have successfully converted the de-duplication query from TrinoSQL to SparkSQL and created a PySpark job for it. However, the second job is incomplete, and the student has not provided a query for it. The student also needs to create tests for the second job.

### Logical Structure
The logical structure of the student's solution is sound for the first job. The student correctly uses a window function to assign row numbers for each group of records with the same `game_id`, `team_id`, and `player_id`, and then selects the first record from each group to de-duplicate the data. However, the second job is incomplete and does not contain any logic.

### Readability
The student's code is generally readable. They have used meaningful variable names and followed Python's naming conventions. However, the student could improve the readability of their code by adding more comments to explain their thought process and the purpose of each function. For example, they could explain why they are using a window function in the `query_1` function and what the `job_1` function does.

### Robustness and Error Handling
The student's code does not include any error handling. They should consider adding try-except blocks to catch potential errors and exceptions. For example, they could catch exceptions when creating or replacing a temporary view or when executing a SQL query.

### Modularity and Reusability
The student's code is modular and reusable. They have separated the logic for creating the SQL query and executing the job into different functions, which makes the code easier to understand and maintain. However, the student could improve the reusability of their code by parameterizing the table name in the `main` function instead of hardcoding it.

### Suggestions for Improvement
1. Complete the second job by providing a SQL query and implementing the logic in the `job_2` function.
2. Create tests for the second job.
3. Add more comments to explain your thought process and the purpose of each function.
4. Add error handling to catch potential errors and exceptions.
5. Parameterize the table name in the `main` function to improve the reusability of your code.

### Grade:
### Query Conversion
**Grade: Needs Improvement**

The student has only converted one query from TrinoSQL to SparkSQL. The conversion is correct, but the second query is missing. The student needs to convert both queries as per the assignment instructions.

### PySpark Jobs
**Grade: Needs Improvement**

The student has created two PySpark jobs, but only one of them is complete. The first job (`job_1`) correctly implements the deduplication logic for the `nba_game_details` table. However, the second job (`job_2`) is incomplete and does not contain any transformation logic. The student needs to complete the second job as per the assignment instructions.

### Unit Tests
**Grade: Satisfactory**

The student has created a unit test for the first job, which correctly tests the deduplication logic. The test uses the `chispa` library to assert DataFrame equality, which is a good practice for testing PySpark jobs. However, the student has not created a unit test for the second job. The student needs to create unit tests for both jobs as per the assignment instructions.

### Code Quality
**Grade: Satisfactory**

The code is well-organized and follows standard Python conventions. The use of docstrings and comments enhances readability. However, the student has left placeholder text ("<YOUR QUERY HERE>") in `job_2.py`, which should be replaced with the actual transformation logic. Also, the second job is incomplete, which affects the overall quality of the code.

### Final Grade
**Needs Improvement**

The student has demonstrated a good understanding of PySpark and SQL, but the assignment is incomplete. The student needs to convert the second query from TrinoSQL to SparkSQL, complete the second PySpark job, and create a unit test for the second job. Once these tasks are completed, the assignment will meet the requirements.

**Comment by julie-scherer** [comment]
### Feedback:
# Feedback

## Coverage of Homework Prompt
The student's solution covers the homework prompt well. They have converted two queries from TrinoSQL to SparkSQL and created new PySpark jobs for these queries. They have also created unit tests for these jobs. However, the student has not provided the converted queries in the homework submission. It would be helpful to include these for completeness and clarity.

## Logical Structure
The student's solution follows a logical structure. The PySpark jobs are well-structured and the SQL queries are written in a clear and understandable manner. The unit tests are also logically structured and test the functionality of the PySpark jobs effectively.

## Readability
The student's code is generally readable. They have used meaningful variable names and the code is well-formatted. However, the student could improve the readability of their code by adding more comments to explain their thought process and the functionality of their code. This would make it easier for others to understand their code.

## Robustness and Error Handling
The student's code does not include explicit error handling. While SparkSQL and PySpark will throw exceptions for many types of errors, it would be beneficial for the student to include explicit error handling in their code to catch and handle potential issues. For example, they could check if the input tables exist before running the jobs, or catch exceptions that might be thrown when executing the SQL queries.

## Modularity and Reusability
The student's code is modular and reusable. The PySpark jobs and SQL queries are defined in separate functions, which makes it easy to reuse them in different contexts. The unit tests are also modular and can be easily adapted to test different functionalities.

## Suggestions for Improvement
- Include the converted SparkSQL queries in the homework submission for completeness and clarity.
- Add more comments to explain the functionality of the code and the thought process behind it.
- Include explicit error handling in the code to catch and handle potential issues.
- Consider using PySpark's DataFrame API for data transformations instead of SQL queries for better performance and readability.

### Grade:
### Query Conversion
**Grade: Satisfactory**
The student has successfully converted the TrinoSQL queries to SparkSQL. However, there are minor issues with the queries. In `job_1.py`, the student has used `{output_table_name}` in the SQL query, but it should be `{BOOTCAMP_TABLE_NAME}`. In `job_2.py`, the student has used `{{output_table_name}}` in the SQL query, which should be `{output_table_name}`.

### PySpark Jobs
**Grade: Satisfactory**
The student has created two PySpark jobs that execute without errors. However, the jobs are not writing to the correct tables. In `job_1.py`, the output is written to `output_table_name` instead of `BOOTCAMP_TABLE_NAME`. In `job_2.py`, the output is written to `output_table_name` instead of `"shabab.user_devices_cumulated"`.

### Unit Tests
**Grade: Satisfactory**
The student has created unit tests for both jobs. The tests cover the main functionalities and use the `assert_df_equality` function from the `chispa` library to compare the actual and expected dataframes. However, the tests could be improved by adding more scenarios and edge cases.

### Code Quality
**Grade: Satisfactory**
The code is well-organized and mostly documented. The student has used type hints and docstrings to improve readability. However, there are minor inconsistencies in style, such as the use of double quotes for strings in some places and single quotes in others.

### Final Grade
**Grade: Satisfactory**
The student has demonstrated a good understanding of SparkSQL and PySpark. The queries have been correctly converted, and the PySpark jobs execute without errors. The unit tests cover the main functionalities of the jobs. However, there are minor issues with the output table names in the jobs and the style consistency in the code. The student is encouraged to pay attention to these details in future assignments.

**Comment by julie-scherer** [comment]
### Feedback:
# Feedback

## Coverage of Homework Prompt
The student's solution covers the homework prompt well. They have converted two queries from TrinoSQL to SparkSQL and created new PySpark jobs for these queries. They have also created unit tests for these jobs. However, the student has not provided the converted queries in the homework submission. It would be helpful to include these for completeness.

## Logical Structure
The student's solution follows a logical structure. The PySpark jobs are well-structured and the SQL queries are written in a clear and understandable manner. The unit tests are also logically structured and test the functionality of the PySpark jobs effectively.

## Readability
The student's code is readable and well-structured. They have used meaningful variable names and included comments to explain the purpose of the code. The SQL queries are also well-formatted and easy to understand. However, the student could improve the readability of their code by adding more comments to explain the logic behind the SQL queries.

## Robustness and Error Handling
The student's code does not include explicit error handling. While SparkSQL will throw an error if the SQL queries fail to execute, it would be beneficial to include try-except blocks to catch any potential errors and provide more informative error messages. Additionally, the student could improve the robustness of their code by adding more comprehensive tests, including edge cases and invalid inputs.

## Modularity and Reusability
The student's code is modular and reusable. The PySpark jobs and SQL queries are defined in separate functions, which can be easily reused in other parts of the code. The unit tests are also modular and can be easily extended to test other functionalities.

## Suggestions for Improvement
- Include the converted SparkSQL queries in the homework submission for completeness.
- Add more comments to explain the logic behind the SQL queries.
- Include try-except blocks to catch any potential errors and provide more informative error messages.
- Add more comprehensive tests, including edge cases and invalid inputs.

### Grade:
### Query Conversion

**Grade: Proficient**

The student has successfully converted two queries from TrinoSQL to SparkSQL. The conversion is accurate and retains the original functionalities of the queries. The student demonstrates a strong understanding of both TrinoSQL and SparkSQL query languages.

### PySpark Jobs

**Grade: Proficient**

The student has created two PySpark jobs that execute without errors. The jobs are well-structured and demonstrate robust Spark capabilities. The student has also made good use of PySpark's DataFrame API and SQL capabilities.

### Unit Tests

**Grade: Proficient**

The student has created comprehensive unit tests for both jobs. The tests cover the main functionalities of the jobs and ensure the integrity of the code. The student has also made good use of the `chispa` library for DataFrame comparisons in Spark, which is a good practice for testing PySpark jobs.

### Code Quality

**Grade: Proficient**

The student's code is well-organized and fully documented. The code follows best practices in modularization, resource use, and maintainability. The use of type hints and docstrings enhances the readability of the code and makes it easier to understand the purpose of the functions and their parameters.

### Final Grade

**Grade: Proficient**

The student has demonstrated a strong understanding of PySpark and SQL, as well as good practices in software development such as unit testing and code documentation. The student's solution is proficient in all areas of the grading rubric.

**Comment by julie-scherer** [comment]
### Feedback:
### Coverage of Homework Prompt
The student's solution covers the homework prompt well. They have successfully converted two queries from TrinoSQL to SparkSQL and created new PySpark jobs for these queries. The student has also created unit tests for these jobs, which use fake input and expected output data. 

### Logical Structure
The student's solution follows a logical structure. The PySpark jobs are well-structured and the SQL queries are correctly converted to SparkSQL. The unit tests are also well-structured and test the functionality of the PySpark jobs effectively.

### Readability
The student's code is highly readable. They have used meaningful variable names and included comments to explain their thought process. The SQL queries are formatted well for readability. The student has also included docstrings for their functions, which provide a clear explanation of the function's purpose and its parameters.

### Robustness and Error Handling
The student's code does not include explicit error handling. However, the nature of the tasks (SQL query conversion and PySpark job creation) does not necessarily require extensive error handling. The unit tests do a good job of testing the functionality of the PySpark jobs and would catch any errors in the SQL queries or the job functions.

### Modularity and Reusability
The student's code is modular and reusable. The PySpark jobs are defined as separate functions, which can be easily reused in other contexts. The SQL queries are defined as separate functions as well, which enhances the modularity of the code. The unit tests are also modular and can be easily adapted to test other similar functions.

### Suggestions for Improvement
The student's solution is already quite good, but there are a few areas where it could be improved:

1. **Error Handling**: While not strictly necessary for this assignment, it would be good practice for the student to include some error handling in their code. For example, they could include try/except blocks around the SQL queries to catch any SQL errors.

2. **Test Coverage**: The student has done a good job of creating unit tests for their PySpark jobs. However, they could improve their test coverage by adding more tests for different edge cases. For example, they could add tests for cases where the input data is empty or contains null values.

3. **Code Comments**: While the student has included some comments in their code, they could improve the readability of their code by adding more comments, especially in the more complex parts of their code. For example, they could add comments explaining the purpose of the different parts of their SQL queries.

### Grade:
### Query Conversion

**Grade: Proficient**

The student has successfully converted two queries from TrinoSQL to SparkSQL. The conversion is accurate, and the student has demonstrated a strong understanding of both TrinoSQL and SparkSQL query languages. The student has also noted the differences between the two SQL dialects, such as the use of `array()` instead of `array[]` in SparkSQL.

### PySpark Jobs

**Grade: Proficient**

The student has created two PySpark jobs that execute the converted queries. The jobs are well-structured and error-free, demonstrating a robust understanding of PySpark. The student has also made good use of PySpark's SQL capabilities by creating temporary views of the input dataframes and executing SQL queries on them.

### Unit Tests

**Grade: Satisfactory**

The student has created unit tests for both PySpark jobs. The tests cover the main functionalities of the jobs and use the `chispa` library to assert dataframe equality. However, the tests could be improved by covering more edge cases and potential error scenarios. For example, the test for `job_2` could be expanded to handle cases where the input dataframes have missing or null values.

### Code Quality

**Grade: Proficient**

The code is well-organized and fully documented. The student has followed best practices in modularization, resource use, and maintainability. The use of namedtuples and f-strings enhances the readability of the code. The student has also demonstrated good error handling by using the `COALESCE` function in the SQL queries to handle null values.

### Final Grade

**Grade: Proficient**

The student has demonstrated a strong understanding of SparkSQL and PySpark, successfully converting TrinoSQL queries to SparkSQL and implementing them in PySpark jobs. The unit tests cover the main functionalities of the jobs, but could be improved by covering more edge cases. The code is well-organized and fully documented, following best practices in modularization and maintainability. Overall, the student has done an excellent job on this assignment.

File: src/jobs/job_2.py [hunk 1]
```diff
@@ -3,18 +3,53 @@
 from pyspark.sql.dataframe import DataFrame
 
 def query_2(output_table_name: str) -> str:
-    query = f"""
-    <YOUR QUERY HERE>
     """
+    Interesting Note: array[] isn't supported in pyspark, had to use array()
```
Comment (HjersmanJ) [inline src/jobs/job_2.py:7]: Good catch

File: src/jobs/job_2.py [hunk 2]
```diff
@@ -3,18 +3,53 @@
...
+    """
+    query = \
+        f"""
+            with
+                yesterday as (
+                    select * from user_devices_cumulated where date = DATE('2022-01-01')
```
Comment (HjersmanJ) [inline src/jobs/job_2.py:13]: Try to keep the where separated to a new line. I'd also make this into a parameter.

File: src/jobs/job_2.py
Comment (HjersmanJ) [inline src/jobs/job_2.py:1]: Good query choices!

File: src/unit_tests/test_spark_jobs.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    Note: see fact-data-modelling query 1
+    """
+
+    game_detail = namedtuple(
+        "game_detail",
+        "game_id team_id team_abbreviation team_city player_id player_name nickname start_position comment min fgm fga fg_pct"
+    )
+
+    # contains duplicates
+    input_data = [
```
Comment (HjersmanJ) [inline src/unit_tests/test_spark_jobs.py:25]: I like this style in showing the first one with column names but omitting it for the rest.

File: src/unit_tests/test_spark_jobs.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+    assert_df_equality(expected_output_df, actual_df, ignore_nullable=True, ignore_row_order=True)
+
+
+def test_job_2_user_devices_cumulated(spark_session):
+    """
+    Test for job2 device cumulated
+    Note: see fact-data-modelling query 3
+
+    TODO: Haven't built spark schemas for this test, save it for future exercise
```
Comment (HjersmanJ) [inline src/unit_tests/test_spark_jobs.py:86]: Yes good use of TODO in comments!

File: src/unit_tests/test_spark_jobs.py
Comment (HjersmanJ) [inline src/unit_tests/test_spark_jobs.py:1]: Would prefer if the test job file was separated by what it is being tested against.

File: src/unit_tests/test_spark_jobs.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    TODO: Note: that this job processes 3 tables, we want to keep a test focussed on single responsibility.
+        However, we could (and should) update our job query to accept kwargs for the other 2,
+        and then write 2 more table focussed unit tests
+    Note: TA (Joe) please don't dock me for not doing the afore mentioned -- I'm maxed out btwn assembly and c programing
```
Comment (HjersmanJ) [inline src/unit_tests/test_spark_jobs.py:90]: Mercy granted

**Review by HjersmanJ — CHANGES_REQUESTED** [review]
You fulfilled the assignment requirements for the most part, and where you didn't, you made a clear roadmap to finish it. I'll have this set as "Request Changes" but will mark you as _Passed_. Put those changes in when you have time along with a proper README. Good job!
